apps/omg/test/omg/state/persistence_test.exs

```diff
@@ -33,172 +34,205 @@ defmodule OMG.State.PersistenceTest do
+
+    assert :ok == exec(create_recovered([{1, 0, 0, alice}], @eth, [{bob, 7}, {alice, 3}]))
+    assert :utxo_not_found == exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 10}]))
   end
 
-  @tag fixtures: [:alice, :state_empty]
-  test "blocks and spends are persisted",
-       %{alice: alice, db_pid: db_pid, state_empty: state} do
+  @tag fixtures: [:alice]
+  test "blocks and spends are persisted", %{alice: alice} do
     tx = create_recovered([{1, 0, 0, alice}], @eth, [{alice, 3}])
 
-    state
-    |> persist_deposit([%{owner: alice.addr, currency: @eth, amount: 20, blknum: 1}], db_pid)
+    [%{owner: alice, currency: @eth, amount: 20, blknum: 1}]
+    |> persist_deposit()
     |> exec(tx)
-    |> persist_form(db_pid)
+    |> persist_form()
+
+    assert {:ok, [hash]} = OMG.DB.block_hashes([@blknum1])
 
-    assert {:ok, [hash]} = OMG.DB.block_hashes([@blknum1], db_pid)
+    :ok = restart_state()
 
-    assert {:ok, [db_block]} = OMG.DB.blocks([hash], db_pid)
+    assert {:ok, [db_block]} = OMG.DB.blocks([hash])
     %Block{number: @blknum1, transactions: [block_tx], hash: ^hash} = Block.from_db_value(db_block)
 
     assert {:ok, tx} == Transaction.Recovered.recover_from(block_tx)
 
     assert {:ok, 1000} ==
-             tx |> Transaction.get_inputs() |> hd() |> InputPointer.Protocol.to_db_key() |> OMG.DB.spent_blknum(db_pid)
+             tx |> Transaction.get_inputs() |> hd() |> InputPointer.Protocol.to_db_key() |> OMG.DB.spent_blknum()
+  end
+
+  @tag fixtures: [:alice]
+  test "exiting utxo is deleted from state", %{alice: alice} do
+    utxo_positions = [
+      Utxo.position(@blknum1, 0, 0),
+      Utxo.position(@blknum1, 0, 1)
+    ]
+
+    [%{owner: alice, currency: @eth, amount: 20, blknum: 1}]
+    |> persist_deposit()
+    |> exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 3}]))
+    |> persist_form()
+    |> persist_exit_utxos(utxo_positions)
+
+    :ok = restart_state()
+
+    assert not OMG.State.utxo_exists?(Utxo.position(@blknum1, 0, 0))
+    assert not OMG.State.utxo_exists?(Utxo.position(@blknum1, 0, 1))
   end
 
-  # mimics `&OMG.State.init/1`
-  defp state_from(db_pid) do
-    {:ok, height_query_result} = OMG.DB.get_single_value(:child_top_block_number, db_pid)
-    {:ok, utxos_query_result} = OMG.DB.utxos(db_pid)
+  @tag fixtures: [:alice]
+  test "cannot spend just exited utxo", %{alice: alice} do
+    :ok = persist_deposit([%{owner: alice, currency: @eth, amount: 20, blknum: 1}])
 
-    {:ok, state} = Core.extract_initial_state(utxos_query_result, height_query_result, @interval)
+    {:ok, _, _} = OMG.State.exit_utxos([Utxo.position(1, 0, 0)])
 
-    state
+    # exit db_updates won't get persisted yet, but alice tries to spent it immediately
+    assert :utxo_not_found == exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 20}]))
+
+    # retry above with empty in-memory utxoset
+    :ok = restart_state()
+
+    {:ok, _, _} = OMG.State.exit_utxos([Utxo.position(1, 0, 0)])
```

Comment:
hm, this shouldn't be necessary and could harm the test, if you weren't persisting

Reply:
IMHO the test is valid as it's race condition possible *persisting exit is done outside of  `OMG.State` process) 
Let say `OMG.State` will have two requests (messages in its inbox)
1. :exit_utxo([alice_utxo])
2. :exec(tx_from_alice_spending_utxo)

So the test shows that even on this edge case there will be no double spend
WDYT?

Reply:
oh right, you're not persisting here, gotcha